### ISSUE TITLE
3.1.0 Regression bug

### ISSUE THREAD

**Issue by wer-mathurin** [Milad-Akarie/auto_route_library#822]
@Milad-Akarie 

There is a bug cause by this line at the line 75 of the routing_controller.dart

` var pendingChildren = const <RouteMatch>[];`

The problem is the **const**
And in the setupRoutes methods you are doing this:
`_routeData.pendingChildren.clear();`

this throw an exception because you can't modify an unmodifiable list.

**Comment by k-s-t-i** [comment]
Having this issue as well.

Here is the stack trace:
```
======== Exception caught by widgets library =======================================================
The following UnsupportedError was thrown building HomeScreen(state: _HomeScreenState#8ad30):
Unsupported operation: Cannot clear an unmodifiable list

The relevant error-causing widget was: 
  HomeScreen HomeScreen:file:///Users/kabirsethi/dev/bitstack/bitstack-app/lib/router/new_router.gr.dart:194:23
When the exception was thrown, this was the stack: 
#0      UnmodifiableListMixin.clear (dart:_internal/list.dart:159:5)
#1      TabsRouter.setupRoutes (package:auto_route/src/router/controller/routing_controller.dart:442:32)
#2      AutoTabsRouterState._setupController (package:auto_route/src/router/widgets/auto_tabs_router.dart:149:18)
#3      AutoTabsRouterState.didChangeDependencies (package:auto_route/src/router/widgets/auto_tabs_router.dart:143:7)
#4      StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:4826:11)
#5      ComponentElement.mount (package:flutter/src/widgets/framework.dart:4638:5)
...     Normal element mounting (38 frames)
#43     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3673:14)
#44     MultiChildRenderObjectElement.inflateWidget (package:flutter/src/widgets/framework.dart:6333:36)
#45     MultiChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:6344:32)
...     Normal element mounting (173 frames)
#218    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3673:14)
#219    MultiChildRenderObjectElement.inflateWidget (package:flutter/src/widgets/framework.dart:6333:36)
#220    Element.updateChild (package:flutter/src/widgets/framework.dart:3425:18)
#221    RenderObjectElement.updateChildren (package:flutter/src/widgets/framework.dart:5758:32)
#222    MultiChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:6356:17)
#223    Element.updateChild (package:flutter/src/widgets/framework.dart:3412:15)
#224    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4690:16)
#225    StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:4840:11)
#226    Element.rebuild (package:flutter/src/widgets/framework.dart:4355:5)
#227    StatefulElement.update (package:flutter/src/widgets/framework.dart:4872:5)
#228    Element.updateChild (package:flutter/src/widgets/framework.dart:3412:15)
#229    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4690:16)
#230    Element.rebuild (package:flutter/src/widgets/framework.dart:4355:5)
#231    ProxyElement.update (package:flutter/src/widgets/framework.dart:5020:5)
#232    Element.updateChild (package:flutter/src/widgets/framework.dart:3412:15)
#233    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4690:16)
#234    Element.rebuild (package:flutter/src/widgets/framework.dart:4355:5)
#235    ProxyElement.update (package:flutter/src/widgets/framework.dart:5020:5)
#236    _InheritedNotifierElement.update (package:flutter/src/widgets/inherited_notifier.dart:181:11)
#237    Element.updateChild (package:flutter/src/widgets/framework.dart:3412:15)
#238    SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:6202:14)
#239    Element.updateChild (package:flutter/src/widgets/framework.dart:3412:15)
#240    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4690:16)
#241    StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:4840:11)
#242    Element.rebuild (package:flutter/src/widgets/framework.dart:4355:5)
#243    StatefulElement.update (package:flutter/src/widgets/framework.dart:4872:5)
#244    Element.updateChild (package:flutter/src/widgets/framework.dart:3412:15)
#245    SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:6202:14)
#246    Element.updateChild (package:flutter/src/widgets/framework.dart:3412:15)
#247    SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:6202:14)
#248    Element.updateChild (package:flutter/src/widgets/framework.dart:3412:15)
#249    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4690:16)
#250    Element.rebuild (package:flutter/src/widgets/framework.dart:4355:5)
#251    ProxyElement.update (package:flutter/src/widgets/framework.dart:5020:5)
#252    Element.updateChild (package:flutter/src/widgets/framework.dart:3412:15)
#253    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4690:16)
#254    StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:4840:11)
#255    Element.rebuild (package:flutter/src/widgets/framework.dart:4355:5)
#256    StatefulElement.update (package:flutter/src/widgets/framework.dart:4872:5)
#257    Element.updateChild (package:flutter/src/widgets/framework.dart:3412:15)
#258    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4690:16)
#259    StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:4840:11)
#260    Element.rebuild (package:flutter/src/widgets/framework.dart:4355:5)
#261    StatefulElement.update (package:flutter/src/widgets/framework.dart:4872:5)
#262    Element.updateChild (package:flutter/src/widgets/framework.dart:3412:15)
#263    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4690:16)
#264    Element.rebuild (package:flutter/src/widgets/framework.dart:4355:5)
#265    ProxyElement.update (package:flutter/src/widgets/framework.dart:5020:5)
#266    Element.updateChild (package:flutter/src/widgets/framework.dart:3412:15)
#267    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4690:16)
#268    Element.rebuild (package:flutter/src/widgets/framework.dart:4355:5)
#269    ProxyElement.update (package:flutter/src/widgets/framework.dart:5020:5)
#270    Element.updateChild (package:flutter/src/widgets/framework.dart:3412:15)
#271    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4690:16)
#272    Element.rebuild (package:flutter/src/widgets/framework.dart:4355:5)
#273    StatelessElement.update (package:flutter/src/widgets/framework.dart:4746:5)
#274    Element.updateChild (package:flutter/src/widgets/framework.dart:3412:15)
#275    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4690:16)
#276    Element.rebuild (package:flutter/src/widgets/framework.dart:4355:5)
#277    ProxyElement.update (package:flutter/src/widgets/framework.dart:5020:5)
#278    Element.updateChild (package:flutter/src/widgets/framework.dart:3412:15)
#279    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4690:16)
#280    Element.rebuild (package:flutter/src/widgets/framework.dart:4355:5)
#281    ProxyElement.update (package:flutter/src/widgets/framework.dart:5020:5)
#282    Element.updateChild (package:flutter/src/widgets/framework.dart:3412:15)
#283    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4690:16)
#284    StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:4840:11)
#285    Element.rebuild (package:flutter/src/widgets/framework.dart:4355:5)
#286    BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2620:33)
#287    WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:882:21)
#288    RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:319:5)
#289    SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1143:15)
#290    SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1080:9)
#291    SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:996:5)
#295    _invoke (dart:ui/hooks.dart:166:10)
#296    PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:270:5)
#297    _drawFrame (dart:ui/hooks.dart:129:31)
(elided 3 frames from dart:async)
```

**Comment by Pipaolo** [comment]
I am also encountering this issue while using the `AutoTabsScaffold` widget.

**Comment by KieranLafferty** [comment]
I believe it's due to this line of code in `routing_controller.dart`
```
RouteData _createRouteData(RouteMatch route, RouteData parent) {
    final mayUpdateController = childControllers[route.key];

    var pendingChildren = const <RouteMatch>[];          <<<<< Remove const so that it can be mutable
    if (mayUpdateController == null && route.hasChildren) {
      pendingChildren = route.children!;
    }
    final routeData = RouteData(
      route: route,
      router: this,
      parent: parent,
      pendingChildren: pendingChildren,
    );

    if (mayUpdateController != null) {
      mayUpdateController._updateRouteDate(routeData);
    }

    return routeData;
  }
```

**Comment by KieranLafferty** [comment]
I'll make a quick pr for this

**Comment by wer-mathurin** [comment]
@Milad-Akarie thanks for the quick turnaround !
FIXED with version 3.1.1
